lib/matplotlib/widgets.py

```diff
@@ -2957,22 +2959,73 @@ def _release(self, event):
 
         self.update()
         self._active_handle = None
+        self._extents_on_press = None
 
         return False
 
     def _onmove(self, event):
         """Motion notify event handler."""
+
+        state = self._state | self._default_state
         # resize an existing shape
         if self._active_handle and self._active_handle != 'C':
             x0, x1, y0, y1 = self._extents_on_press
-            if self._active_handle in ['E', 'W'] + self._corner_order:
-                x1 = event.xdata
-            if self._active_handle in ['N', 'S'] + self._corner_order:
-                y1 = event.ydata
+            sizepress = [x1 - x0, y1 - y0]
+            center = [x0 + sizepress[0] / 2, y0 + sizepress[1] / 2]
+            dx = event.xdata - self._eventpress.xdata
+            dy = event.ydata - self._eventpress.ydata
+
+            # change sign of relative changes to simplify calculation
+            # Switch variables so that only x1 and/or y1 are updated on move
+            x_factor = y_factor = 1
+            if 'W' in self._active_handle:
+                x_factor *= -1
+                dx *= x_factor
+                x0 = x1
+            if 'S' in self._active_handle:
+                y_factor *= -1
+                dy *= y_factor
+                y0 = y1
+
+            # from center
+            if 'center' in state:
+                if 'square' in state:
+                    if self._active_handle in ['E', 'W']:
+                        # using E, W handle we need to update dy accordingly
+                        dy = dx
+                    elif self._active_handle in ['S', 'N']:
+                        # using S, N handle, we need to update dx accordingly
+                        dx = dy
+                    else:
+                        dx = dy = max(dx, dy, key=abs)
+
+                dw = sizepress[0] / 2 + dx
+                dh = sizepress[1] / 2 + dy
```

Comment:
Are these actually changes in width/height? Are they the new half{width,height} instead? If so I think they should have a name other than d{w,h}, as the d implies a change.

Reply:
Yes, indeed, I renamed them to `hw` and `hh` and added a comment.

Reply:
Thanks - now you mention it, I've seen `hw` and `hh` used elsewhere, so that's good!